integrations: Prevent duplicate PR review messages.

PieterCK opened the pull request:
This PR removes duplicative messages from payloads related to starting a PR review with code comments. This is done by filtering out an identical payload similar to `pull_request_review.json` but with "edited" as the "action" value and contains an additional empty "changes" dict which is documented in the new `pull_request_review_duplicate_payload.json` fixture. Refer to this diff to see these differences: https://www.diffchecker.com/dlrsAXQc/

The duplicate message from #26145 has been successfully reproduced. The result is in this [discussion](https://chat.zulip.org/#narrow/stream/127-integrations/topic/github.20bot/near/1886176)

Fixes #26145.

**Screenshots and screen captures:**
### GitHub PR review:
![image](https://github.com/user-attachments/assets/55b123d6-3e46-457f-ad8b-b9810a08c089)

### Before:
![image](https://github.com/user-attachments/assets/eb3dbdeb-b488-4b50-9abe-21b8afbd8cd7)

### Fixed:
![image](https://github.com/user-attachments/assets/09261cc2-7f54-4eb4-b0a8-de7ad9981c63)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK commented:
@zulipbot add "buddy review" 

@kennethnrk - Please review this one!

PieterCK commented on zerver/webhooks/github/tests.py:
```diff
@@ -479,6 +479,18 @@ def test_ignored_pull_request_actions(self) -> None:
+            payload = orjson.dumps(data).decode()
+            self.verify_post_is_ignored(payload, "pull_request_review")
+
+        self.check_webhook("pull_request_review_duplicate_payload", "", "", expect_noop=True)
```
> **sbansal1999** : We can probably add a fixture for this test. Feel free to use the one that was discussed [here](https://chat.zulip.org/#narrow/stream/127-integrations/topic/github.20bot/near/1600140) or you can generate a new one as well.

additionally, I've extended the `test_ignored_pull_request_review_actions` test to use the `check_webhook` function with the duplicate payload. Do we need to keep the test logic using `verify_post_is_ignored` here? it feels redundant but also consistent with all other ignored related tests.

sbansal1999 replied:
Since we are not completely ignoring all the `edited` events, I feel that `ignored_actions` array gives a wrong indication. Something like this would be a better solution, in my opinion:

```python
   def test_pull_request_review_edited_with_empty_changes_ignore(self) -> None:
        payload = self.get_body("pull_request_review__edited_empty_changes")
        self.verify_post_is_ignored(payload, "pull_request_review")
```

sbansal1999 replied:
Looks like one comment got deleted smh,

Since we are not completely ignoring all the `edited` events, I feel that `ignored_actions` array gives a wrong indication. Something like [this](https://github.com/zulip/zulip/blob/main/zerver/webhooks/github/tests.py#L465-L467) would be a better solution, in my opinion.

PieterCK replied:
Ah, okay, that makes more sense.

PieterCK commented on zerver/webhooks/github/view.py:
```diff
@@ -970,6 +970,12 @@ def get_zulip_event_name(
+        action = payload["action"].tame(check_string)
+        changes = payload.get("changes", {})
+        if action == "edited" and len(changes) == 0:
+            return None
```
> **sbansal1999**: The comment is probably not required here, as the code is pretty much self explanatory.

removed comment

kennethnrk commented:
LGTM, payloads with "edited" as the action do not create a duplicate "PR review" messages.

@zulipbot remove "buddy review"

PieterCK commented:
@zulipbot add "mentor review" 

@sbansal1999 Please review this!

sbansal1999 commented on zerver/webhooks/github/view.py:
```diff
@@ -970,6 +970,12 @@ def get_zulip_event_name(
             return None
+    elif header_event == "pull_request_review":
+        action = payload["action"].tame(check_string)
+        changes = payload.get("changes", {})
+        if action == "edited" and len(changes) == 0:
```
Better to abstract this check through a function.

sbansal1999 reviewed (commented):
@PieterCK Thanks for working on this one.

Left some comments, shouldn't take much time to fix them. 

Feel free to comment if you have any questions or want to discuss anything.

PieterCK commented:
@sbansal1999 - Thanks for the review! I've addressed your comments. Let me know if the new changes work for you and if there are any other concerns you'd like to point out.

PieterCK commented:
Update: addressed review

sbansal1999 commented on zerver/webhooks/github/tests.py:
```diff
@@ -479,6 +479,10 @@ def test_ignored_pull_request_actions(self) -> None:
             payload = orjson.dumps(data).decode()
             self.verify_post_is_ignored(payload, "pull_request")
 
+    def test_ignored_pull_request_review_actions(self) -> None:
+        payload = self.get_body("pull_request_review__edited_empty_changes")
+        self.verify_post_is_ignored(payload, "pull_request_review")
+
```
Since this test only checks the empty changes edited actions and not all ignored pull request review actions. It's better to name it as `test_pull_request_review_edited_empty_changes_ignore`.

PieterCK commented:
Updated test function name

sbansal1999 commented:
LGTM. Marking this for maintainer review. @laurynmm 

@zulipbot add "maintainer review"

sbansal1999 commented on zerver/webhooks/github/fixtures/pull_request_review_duplicate_payload.json:
```diff
@@ -0,0 +1,515 @@
+{
```
Also, you can update the commit message accordingly to have the new file name. Ideally, there is no need to mention which file has been added in the commit message, as this can be seen by simply checking out that commit.

PieterCK commented:
Thanks again for the review!

PieterCK commented:
@zulipbot remove "mentor review"

laurynmm reviewed (commented):
@PieterCK - These changes look good! My one comment is that the commit message needs to be rewritten to actually reflect what's happening in these changes.

A few notes on how I'd revise the commit message text:
- The commit summary should mention that this is specifically for the GitHub integration and not for integrations in general. Something like `integrations: Prevent duplicate GitHub pull request review messages.`
- Let's avoid abbreviations like PR if possible and use "pull request" instead.
- The final sentence is incorrect as we're ignoring the "edited" payload, so that should be changed.
- I'd simplify the description of the problem as we have the issue description and a CZO conversation that goes over all of that in more depth. I'd revise those two paragraphs to be something like ...

```
GitHub sends two almost identical payloads when a pull request is
reviewed, which results in ...

The payloads have different ...

We now ignore the ...
```

Let me know if you have any questions about my notes above.

laurynmm commented:
@PieterCK - Thanks for updating the commit message. This looks good to me now, so I'm marking for integration review.

timabbott commented:
Nice, added a comment on this corner case in the code for posterity, and merged, thanks @PieterCK!

Tagging as a backport candidate; we'll want to include in 9.1 since it's a nice fix to an important integration.

timabbott commented:
Backported in https://github.com/zulip/zulip/pull/31216.
